codex/sales/reservations.nim

```diff
@@ -0,0 +1,349 @@
+## Nim-Codex
+## Copyright (c) 2022 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import std/typetraits
+
+import pkg/chronos
+import pkg/chronicles
+import pkg/upraises
+import pkg/json_serialization
+import pkg/json_serialization/std/options
+import pkg/stint
+import pkg/stew/byteutils
+import pkg/nimcrypto
+import pkg/questionable
+import pkg/questionable/results
+
+push: {.upraises: [].}
+
+import pkg/datastore
+import ../stores
+import ../contracts/requests
+
+export requests
+
+logScope:
+  topics = "reservations"
+
+type
+  AvailabilityId* = distinct array[32, byte]
+  Availability* = object
+    id*: AvailabilityId
+    size*: UInt256
+    duration*: UInt256
+    minPrice*: UInt256
+    used*: bool
+  Reservations* = ref object
+    repo: RepoStore
+  GetNext* = proc(): Future[?Availability] {.upraises: [], gcsafe, closure.}
+  AvailabilityIter* = ref object
+    finished*: bool
+    next*: GetNext
+  AvailabilityError* = object of CodexError
+  AvailabilityAlreadyExistsError* = object of AvailabilityError
+  AvailabilityReserveFailedError* = object of AvailabilityError
+  AvailabilityReleaseFailedError* = object of AvailabilityError
+  AvailabilityDeleteFailedError* = object of AvailabilityError
+  AvailabilityGetFailedError* = object of AvailabilityError
+  AvailabilityUpdateFailedError* = object of AvailabilityError
+
+const
+  SalesKey = (CodexMetaKey / "sales").tryGet # TODO: move to sales module
+  ReservationsKey = (SalesKey / "reservations").tryGet
+
+proc new*(
+  T: type Reservations,
+  repo: RepoStore): Reservations =
+
+  T(repo: repo)
+
+proc init*(
+  _: type Availability,
+  size: UInt256,
+  duration: UInt256,
+  minPrice: UInt256): Availability =
+
+  var id: array[32, byte]
+  doAssert randomBytes(id) == 32
+  Availability(id: AvailabilityId(id), size: size, duration: duration, minPrice: minPrice)
+
+func toArray*(id: AvailabilityId): array[32, byte] =
+  array[32, byte](id)
+
+proc `==`*(x, y: AvailabilityId): bool {.borrow.}
+proc `==`*(x, y: Availability): bool =
+  x.id == y.id and
+  x.size == y.size and
+  x.duration == y.duration and
+  x.minPrice == y.minPrice
+
+proc `$`*(id: AvailabilityId): string = id.toArray.toHex
+
+proc toErr[E1: ref CatchableError, E2: AvailabilityError](
+  e1: E1,
+  _: type E2,
+  msg: string = "see inner exception"): ref E2 =
```

Comment:
I don't think that Nim by default prints the inner exception, so perhaps it's better to use the original msg?

```suggestion
  msg: string = e1.msg): ref E2 =
```

Reply:
Good call.